builtin/logical/database/path_creds_create.go

```diff
@@ -54,24 +54,10 @@ func (b *databaseBackend) pathCredsCreateRead() framework.OperationFunc {
 			return nil, logical.ErrPermissionDenied
 		}
 
-		// Grab the read lock
-		b.RLock()
-		unlockFunc := b.RUnlock
-
 		// Get the Database object
-		db, ok := b.getDBObj(role.DBName)
-		if !ok {
-			// Upgrade lock
-			b.RUnlock()
-			b.Lock()
-			unlockFunc = b.Unlock
-
-			// Create a new DB object
-			db, err = b.createDBObj(ctx, req.Storage, role.DBName)
-			if err != nil {
-				unlockFunc()
-				return nil, fmt.Errorf("cound not retrieve db with name: %s, got error: %s", role.DBName, err)
-			}
+		db, err := b.GetConnection(ctx, req.Storage, role.DBName)
```

Comment:
So we no longer are holding the lock while using a db object. What if another thread closes this object?

Reply:
I guess I misunderstood what the lock was protecting.  I assumed we were protecting concurrent access to the connections map.  Looking at that `pg` lib, it looks like inflight may already be protected by a lock, so it is probably fine and really up to the library author to no yank out connections.

Another option might be to defer a database close for some period of time, maybe 5 seconds.  I was already thinking about this for other reasons, but we would want to ensure that the proper connection is closed.  I think we need to tag the connection with a uuid, so when we close it we know it is the correct one.

Reply:
I think the lock should protect the access to the connection map and from the plugin client killing the plugin process out from underneath us.

Reply:
I decided there should be different locks for these two concerns.  I wrapped each db plugin connection and added a lock to that.  I think this should address the state changes of the connection map and killing of a plugin instance while requests are in flight.